### DISCUSSION TITLE
Stop flattening The Hive's output into 3 categories of suggestive material

### DISCUSSION THREAD

**Discussion by nuklearfiziks** [bluesky-social/atproto#1466] [Protocol (atproto)]
Hi! 💚 

I run the "🔞 Suggestive", "🔥🔞 Hot Suggestive" and "Described Suggestive" feeds. 

Please, dear god please, stop [flattening the results from The Hive into like 3 classes of sexualised content](https://github.com/bluesky-social/atproto/blob/9a824aeae63eacde2f3b10eb276a84e16f92d6ff/packages/pds/src/labeler/hive.ts#L78-L146).

What is even the design decision behind this? What does it accomplish, beyond limiting the ability of users to control what content they see?

Not only that, if you're going to moderate content decisions algorithmically, be transparent about what the algorithm is returning! Just a few weeks ago we had users running their images through The Hive's demo tool in order to compare them against what Bluesky is doing with its labeller, *simply because this is so opaque*. Not only would returning the raw labeller data *in addition to* the summarised labels make it easier for people like me to make better feeds, but it would also improve user trust because we'd all have a much better idea what the third-party company you're shipping all of our images to is doing with them.

Please! Return all the dang labels (and their scores)!

-Andi.

**Comment by GreenFootballs** [comment]
An even better idea would be to let users add their own labels if necessary.

**Reply by nuklearfiziks** [reply]
I think that's an aspect of composible moderation intended for further down the road; what I'm asking for in the meantime is much less dramatic, would have a minimal impact on existing infrastructure and provide a lot of much-needed transparency about moderation decisions being ultimately made via a proprietary machine learning model.

**Reply by DavidBuchanan314** [reply]
btw, self-labelling (in the 3 categories mentioned above) is already implemented and merged, it's just not deployed to prod yet.

**Reply by Zero3K** [reply]
Why hasn't it been deployed to production yet? Does it still need more work done on it?

**Reply by intrnl** [reply]
Self-labeling is already there, people has been using it to create posts using the !warn label which puts your post in a spoiler like sort of thing, the app has been versioned and that contains the toggles to self-label suggestive material but I assume they're waiting for App Store reviews or whatnot

**Reply by Zero3K** [reply]
Huh?

**Reply by pfrazee** [reply]
Yes, it's caught up in an unusually long appstore review

**Comment by DavidBuchanan314** [comment]
The way I see it, "Hive" is an implementation detail, and other content labellers are inevitably going to have their own internal representations. The labelling lexicons are an attempt to make a common language between potentially competing labelling solutions.

*However*, I can also understand the desire for more transparency. I know I've been running images through Hive myself, to attempt to understand a classification.

I think a potential solution would be to have some kind of free-form "explanation" or "justification" field, as part of the label - which as you suggest, could contain raw output from Hive.

A couple of related thoughts:

1. I would personally not want to use raw hive classifications as a basis for building custom feeds, since you're, in effect, relying on an implementation detail, and such things could result in ecosystem-scale vendor lock-in in the long run. But of course, you're free to do whatever you want.

2. A free-text "explanation"/"justification"/"details" field could become a vector for abuse, in the future, when 3rd party labeling services are a thing

**Reply by nuklearfiziks** [reply]
This is really helpful context; thank you. Point taken about The Hive as an implementation detail. That said, if the desire is to create a shared vocabulary of content labels, I'm not sure the current distinction between "sexual," "porn," and "nudity" (afaict "underwear" is unused?) is particularly illuminating — I just launched 4 new variants of "Suggestive" that are exclusively one of the four explicit content labels and honestly they're pretty indistinguishable.

I do think a "reason" field would be useful, particularly when a lot of the decisions made by ML-driven systems can come off as opaque and arbitrary to users.

**Reply by pfrazee** [reply]
There are a couple of overlapping questions in here.

1. what is the logic of the labels vocabulary broadly?
2. what is the logic of the labels as they currently exist?
3. can we introduce a `reason` or similar to elaborate on labels?

**The broad logic**. Labels are currently a well-defined vocabulary (which you can find here: https://github.com/bluesky-social/atproto/blob/main/packages/api/docs/labels.md). There are two directions you can go with a vocabulary, well-defined (aka in spec docs) or open (like how hashtags work; a "folksonomy"). I've leaned toward well-defined because you want everybody to agree one what they mean and how to interpret them. The presence of a label doesn't _automatically_ mean that the content will be masked; it depends on the user settings. If new labels are cropping up all the time -- and with differing interpretations -- then users won't be able to configure them.

If you're aiming for the general kind of warning that covers the content but which doesn't filter the post from feeds, use the `!warning` label. That's a special label designed for that.

**The logic of sexual/nudity/porn.** This is not exactly exact science so I won't say we've for sure nailed it, but we're trying to delineate between hornyposting (sexual) non-sexual nudity eg in art (nudity) and then full-on bedroom activity (porn). Our _goal_ is to avoid an overly broad NSFW category that catches too much; the softer categories (sexual and nudity) are meant to give adults a finer control over what they see. In practice it's subjective enough that it can sometimes seem arbitrary, so we'll see how well this plays out over time.

**Adding "reason" fields.** I think there are some interesting things we can do here for sure. We should talk about the abuse vectors -- off the top of my head I'm unsure what the risks are -- but I'm sure some folks would love to be able to describe what's on offer, so to speak
